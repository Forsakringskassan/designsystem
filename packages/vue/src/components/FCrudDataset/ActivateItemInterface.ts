import { inject } from "vue";

/**
 * @public
 */
export type ActivateItemCallback<T> = (item: T) => void;

/**
 * @public
 */
export interface ActivateItemInterface<T> {
    // After an item is added it should be active
    registerCallbackAfterItemAdd(callback: ActivateItemCallback<T>): void;
    // When an item is deleted the item above it should be set to active and have focus.
    registerCallbackBeforeItemDelete(callback: ActivateItemCallback<T>): void;
}

/**
 * @public
 */
export function ActivateItemInjected<T>(): ActivateItemInterface<T> {
    return {
        registerCallbackAfterItemAdd: inject(
            "registerCallbackAfterItemAdd",
            () => undefined,
        ) as (callback: ActivateItemCallback<T>) => void,
        registerCallbackBeforeItemDelete: inject(
            "registerCallbackBeforeItemDelete",
            () => undefined,
        ) as (callback: ActivateItemCallback<T>) => void,
    };
}
