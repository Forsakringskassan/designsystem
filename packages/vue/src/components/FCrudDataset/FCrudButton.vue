<template>
    <button type="button" class="button button--small button--tertiary" @click="executeAction">
        <f-icon v-if="icon" class="button__icon" :name="iconName"></f-icon>
        <span v-if="!label" class="sr-only">
            <slot> {{ buttonText }} </slot>
        </span>
        <slot v-if="label"> {{ buttonText }} </slot>
    </button>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { FIcon } from "../FIcon";
import { TranslationMixin } from "../../plugins";
import { FCrudDatasetInjected, FCrudDatasetInterface } from "./FCrudDatasetInterface";

export default defineComponent({
    name: "FCrudButton",
    components: { FIcon },
    mixins: [TranslationMixin],
    props: {
        action: {
            type: String,
            required: true,
            validator(value: string) {
                return ["delete", "modify"].includes(value);
            },
        },
        icon: {
            type: Boolean,
            default: false,
        },
        item: {
            type: Object,
            required: true,
        },
        label: {
            type: Boolean,
            default: false,
        },
    },
    setup(): FCrudDatasetInterface {
        return FCrudDatasetInjected();
    },
    computed: {
        iconName(): string {
            if (this.action === "delete") {
                return "trashcan";
            } else {
                return "pen";
            }
        },
        buttonText(): string {
            if (this.action === "delete") {
                return this.$t("fkui.crud-button.delete", "Ta bort");
            } else {
                return this.$t("fkui.crud-button.modify", "Ã„ndra");
            }
        },
    },
    methods: {
        executeAction(): void {
            if (this.action === "delete") {
                this.delete(this.item);
            } else {
                this.modify(this.item);
            }
        },
    },
});
</script>
