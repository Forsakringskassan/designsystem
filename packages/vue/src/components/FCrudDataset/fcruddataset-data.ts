import { type ListArray, type ListItem } from "../../types";
import { Operation } from "./operation";

/**
 * @public
 */
export interface FCrudDatasetData {
    result: ListArray;
    Operation: typeof Operation;
    operation: Operation;
    item: null | ListItem;
    originalItemToUpdate: null | ListItem;
    isFormModalOpen: boolean;
    isConfirmModalOpen: boolean;
    callbackAfterItemAdd(item: ListItem): void;
    callbackBeforeItemDelete(item: ListItem): void;
}
