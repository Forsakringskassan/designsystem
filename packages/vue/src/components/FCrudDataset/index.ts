export { default as FCrudDataset } from "./FCrudDataset.vue";
export { default as FCrudButton } from "./FCrudButton.vue";
export {
    type ActivateItemCallback,
    type ActivateItemInterface,
    ActivateItemInjected,
} from "./ActivateItemInterface";
export { type FCrudDatasetInterface } from "./FCrudDatasetInterface";
export { Operation } from "./operation";
