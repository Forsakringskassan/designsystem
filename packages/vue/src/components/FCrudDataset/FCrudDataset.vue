<template>
    <div class="crud-dataset">
        <!--
@slot Slot for displaying the data.
    -->
        <slot></slot>
        <div v-if="hasAddSlot">
            <button
                data-test="f-crud-dataset-add-button"
                type="button"
                class="button button--tertiary crud-dataset__add-button"
                @click="createItem()"
            >
                <f-icon class="button__icon" name="plus" />
                <!--
@slot Slot for changing the text in "Add new" button`
                    -->
                <slot name="add-button">{{ $t("fkui.crud-dataset.button.add", "Lägg till ny") }}</slot>
            </button>
        </div>

        <f-form-modal
            :is-open="isFormModalOpen"
            :aria-close-text="$t('fkui.crud-dataset.modal.close', 'Stäng')"
            :use-error-list="false"
            :before-submit="beforeSubmit"
            :before-validation="beforeValidation"
            :on-cancel="onCancel"
            @close="onModalClose"
            @cancel="onCancel"
            @submit="onFormModalSubmit"
        >
            <template #header>{{ formModalHeader }}</template>
            <template #input-text-fields>
                <!--
@slot Slot for inputs when creating a new item.
The new item is available through `v-slot="{ item }"`
If any data in the item should be set not by the user the prop beforeCreate can be used to set that data.
                    -->
                <slot v-if="operation === Operation.ADD" name="add" v-bind="{ item: item }" />
                <!--
@slot Slot for inputs when modifying an item.
The item being modified is available through `v-slot="{ item }"`
                    -->
                <slot v-if="operation === Operation.MODIFY" name="modify" v-bind="{ item: item }" />
            </template>
            <template #submit-button-text>{{ confirmButtonText }}</template>
            <template #cancel-button-text>{{ cancelButtonText }}</template>
        </f-form-modal>

        <f-confirm-modal
            :is-open="isConfirmModalOpen"
            :buttons="confirmDeleteButtons"
            @confirm="onDeleteConfirm"
            @close="onDeleteClose"
        >
            <template #heading>
                {{ deleteModalHeader }}
            </template>

            <template #content>
                <!--
@slot Slot for displaying a warning message before an item is deleted.
The item being deleted is available through `v-slot="{ item }"`
                    -->
                <slot name="delete" v-bind="{ item: item }" />
            </template>
        </f-confirm-modal>
    </div>
</template>

<script lang="ts">
import { defineComponent, type PropType } from "vue";
import { deepClone, alertScreenReader, TranslationService } from "@fkui/logic";

import { FIcon } from "../FIcon";
import { FFormModal, FConfirmModal } from "../FModal";
import { FModalButtonDescriptor } from "../FModal/modal-button";
import { type ListItem, type ListArray } from "../../types";
import { TranslationMixin } from "../../plugins";
import { type FValidationFormCallback } from "../FValidationForm";
import { FCrudDatasetInterface } from "./FCrudDatasetInterface";
import { ActivateItemInterface } from "./ActivateItemInterface";
import { type FCrudDatasetData } from "./fcruddataset-data";
import { Operation } from "./operation";

export default defineComponent({
    name: "FCrudDataset",
    components: { FFormModal, FConfirmModal, FIcon },
    mixins: [TranslationMixin],
    provide(): FCrudDatasetInterface & ActivateItemInterface {
        return {
            delete: (item: ListItem) => {
                this.deleteItem(item);
            },
            modify: (item: ListItem) => {
                this.updateItem(item);
            },
            registerCallbackAfterItemAdd: (callback: (item: ListItem) => void) => {
                this.callbackAfterItemAdd = callback;
            },
            registerCallbackBeforeItemDelete: (callback: (item: ListItem) => void) => {
                this.callbackBeforeItemDelete = callback;
            },
        };
    },
    props: {
        /**
         * The list of items that should be deleted, modified or added to.
         * If the prop is not set an empty array will be used.
         * @model
         */
        modelValue: {
            type: Array as PropType<ListArray<ListItem>>,
            required: false,
            default: () => [],
        },
        /**
         * A function that returns an item to the #add template. Can be used to populate data that the user should not input themself e.g. an id.
         * Or to give the user suggestions for inputs. If the prop is not used an empty item will be returned.
         */
        beforeCreate: {
            type: Function as PropType<(() => ListItem) | undefined>,
            required: false,
            default: undefined,
        },
        /**
         * If `true` the primary button in the modals will be placed to the right side instead of to the left.
         */
        primaryButtonRight: {
            type: Boolean,
            default: false,
        },
        /**
         * If given, this function is called before the [[submit]] event is emitted.
         * See <f-validation-form> `beforeSubmit` props for more info.
         */
        beforeSubmit: {
            type: Function as PropType<FValidationFormCallback>,
            required: false,
            default(): void {
                /* do nothing */
            },
        },
        /**
         * If given, this function is called before the form data is validated and the [[submit]] event is emitted.
         * See <f-validation-form> `beforeValidation` props for more info.
         */
        beforeValidation: {
            type: Function as PropType<FValidationFormCallback>,
            required: false,
            default(): void {
                /* do nothing */
            },
        },
        /**
         * If given, this function is called after the modal has been closed.
         */
        onCancel: {
            type: Function as PropType<() => void>,
            required: false,
            default() {
                return undefined;
            },
        },

        /**
         * Property for changing the "add new" modal heading
         */
        addNewModalHeader: {
            type: String,
            required: false,
            default: TranslationService.provider.translate("fkui.crud-dataset.modal.header.add", "Lägg till rad"),
        },
        /**
         * Property for changing the "modify" modal heading
         */
        modifyModalHeader: {
            type: String,
            required: false,
            default: TranslationService.provider.translate("fkui.crud-dataset.modal.header.modify", "Ändra rad"),
        },
        /**
         * Property for changing the "delete" modal heading
         */
        deleteModalHeader: {
            type: String,
            required: false,
            default: TranslationService.provider.translate(
                "fkui.crud-dataset.modal.header.delete",
                "Är du säker på att du vill ta bort raden?",
            ),
        },
    },
    emits: ["change", "created", "deleted", "updated", "update:modelValue"],
    data(): FCrudDatasetData {
        return {
            result: [],
            Operation,
            operation: Operation.NONE,
            item: null,
            originalItemToUpdate: null,
            isFormModalOpen: false,
            isConfirmModalOpen: false,
            callbackAfterItemAdd() {
                // dummy method when not provided by underlying component
            },
            callbackBeforeItemDelete() {
                // dummy method when not provided by underlying component
            },
        };
    },
    computed: {
        confirmButtonText(): string {
            return this.operation === Operation.ADD
                ? this.$t("fkui.crud-dataset.modal.confirm.add", "Lägg till")
                : this.$t("fkui.crud-dataset.modal.confirm.modify", "Spara");
        },
        cancelButtonText(): string {
            return this.operation === Operation.ADD
                ? this.$t("fkui.crud-dataset.modal.cancel.add", "Avbryt")
                : this.$t("fkui.crud-dataset.modal.cancel.modify", "Avbryt");
        },
        confirmDeleteButtons(): FModalButtonDescriptor[] {
            return [
                {
                    label: this.$t("fkui.crud-dataset.modal.confirm.delete", "Ja, ta bort"),
                    type: "primary",
                    event: "confirm",
                },
                { label: this.$t("fkui.crud-dataset.modal.cancel.delete", "Nej, avbryt"), type: "secondary" },
            ];
        },
        hasAddSlot(): boolean {
            return Boolean(this.$slots.add);
        },
        hasDeleteSlot(): boolean {
            return Boolean(this.$slots.delete);
        },
        hasModifySlot(): boolean {
            return Boolean(this.$slots.modify);
        },
        formModalHeader(): string {
            return this.operation === Operation.ADD ? this.addNewModalHeader : this.modifyModalHeader;
        },
    },
    watch: {
        modelValue: {
            immediate: true,
            deep: true,
            handler(data: ListItem[]) {
                this.result = [...data];
            },
        },
    },
    mounted(): void {
        if (!this.hasAddSlot && !this.hasDeleteSlot && !this.hasModifySlot) {
            throw Error("Atleast one template of the following must be defined. #add, #delete or #modify");
        }
    },
    methods: {
        createItem(): void {
            if (!this.hasAddSlot) {
                throw Error("No template is defined for #add");
            }
            this.operation = Operation.ADD;
            this.item = this.beforeCreate ? this.beforeCreate() : {};
            this.isFormModalOpen = true;
        },
        deleteItem(item: ListItem) {
            if (!this.hasDeleteSlot) {
                throw Error("No template is defined for #delete");
            }
            this.operation = Operation.DELETE;
            this.item = item;
            this.isConfirmModalOpen = true;
        },
        onDeleteConfirm(): void {
            if (!this.item) {
                return;
            }
            this.callbackBeforeItemDelete(this.item);
            this.result = this.result.filter((item: ListItem) => item !== this.item);

            /**
             * Emitted when an item is deleted.
             * @event deleted
             * @param item - the deleted item.
             */
            this.$emit("deleted", this.item);

            /**
             * V-model event.
             * @event update:modelValue
             */
            this.$emit("update:modelValue", this.result);

            /**
             * Vue2 v-model event.
             * @deprecated
             * @event change
             */
            this.$emit("change", this.result);

            alertScreenReader(this.$t("fkui.crud-dataset.aria-live.delete", "Raden har tagits bort"), {
                assertive: true,
            });
        },
        onDeleteClose(e: { reason: string }) {
            this.onModalClose();
            if (e.reason === "close" && this.onCancel) {
                this.onCancel();
            }
        },

        onModalClose(): void {
            this.isFormModalOpen = false;
            this.isConfirmModalOpen = false;
        },
        onFormModalSubmit(): void {
            if (!this.item) {
                return;
            }
            if (this.operation === Operation.ADD) {
                this.result.push(this.item);
                /**
                 * Emitted when an item is added.
                 * @event created
                 * @param item - the added item.
                 */
                this.$emit("created", this.item);
                this.$emit("update:modelValue", this.result);
                this.$emit("change", this.result);

                this.callbackAfterItemAdd(this.item);
                alertScreenReader(this.$t("fkui.crud-dataset.aria-live.add", "En rad har lagts till"), {
                    assertive: true,
                });
            } else if (this.operation === Operation.MODIFY) {
                if (this.originalItemToUpdate) {
                    Object.assign(this.originalItemToUpdate, this.item);
                } else {
                    this.originalItemToUpdate = this.item;
                }
                /**
                 * Emitted when an item is updated.
                 * @event updated
                 * @param item - the updated item.
                 */
                this.$emit("updated", this.originalItemToUpdate);
                this.$emit("update:modelValue", this.result);
                this.$emit("change", this.result);

                alertScreenReader(this.$t("fkui.crud-dataset.aria-live.modify", "Raden har ändrats"), {
                    assertive: true,
                });
            }
            this.isFormModalOpen = false;
        },
        updateItem(item: ListItem): void {
            if (!this.hasModifySlot) {
                throw Error("No template is defined for #modify");
            }
            this.operation = Operation.MODIFY;
            this.originalItemToUpdate = item;
            this.item = deepClone(item);
            this.isFormModalOpen = true;
        },
    },
});
</script>
