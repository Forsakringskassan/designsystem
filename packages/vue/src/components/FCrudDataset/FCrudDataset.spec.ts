import "html-validate/jest";
import { mount, VueWrapper } from "@vue/test-utils";
import { type PropType, defineComponent } from "vue";
import flushPromises from "flush-promises";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import logic from "@fkui/logic";
import { ListItem } from "../../types";
import { FFormModal, FFormModalAction, FModal } from "../FModal";
import { FTextField } from "../FTextField";
import { type FValidationFormCallback } from "../FValidationForm";
import { ValidationPlugin } from "../../plugins";
import FCrudDataset from "./FCrudDataset.vue";
import FCrudButton from "./FCrudButton.vue";

jest.useFakeTimers();

afterEach(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
});

const ADD_TEMPLATE = /* HTML */ `
    <template #add="{ item }">
        <span> add </span>
        <input id="id-input" type="number" v-model.number="item.id" />
        <input id="name-input" type="text" v-model="item.name" />
    </template>
`;

const ADD_TEMPLATE_REQUIRED = /* HTML */ `
    <template #add="{ item }">
        <span> add </span>
        <f-text-field
            v-validation.required
            id="id-input-required"
            type="number"
            v-model.number="item.id"
        ></f-text-field>
        <f-text-field
            v-validation.required
            id="name-input-required"
            type="text"
            v-model="item.name"
        ></f-text-field>
        <input />
    </template>
`;

const MODIFY_TEMPLATE = /* HTML */ `
    <template #modify="{ item }">
        <span> modify </span>
        <input id="name-input" type="text" v-model="item.name" />
        <input id="child-name-input" type="text" v-model="item.child.name" />
    </template>
`;

const DELETE_TEMPLATE = /* HTML */ `
    <template #delete="{ item }">
        <span> delete </span>
        {{ item.name }}?
    </template>
`;

const TestComponent = defineComponent({
    name: "TestComponent",
    components: { FCrudDataset, FCrudButton, FModal, FTextField },
    props: {
        beforeSubmit: {
            type: Function as PropType<FValidationFormCallback>,
            required: false,
            default(): void {
                /* do nothing */
            },
        },
        beforeValidation: {
            type: Function as PropType<FValidationFormCallback>,
            required: false,
            default(): void {
                /* do nothing */
            },
        },
        onCancel: {
            type: Function as PropType<FValidationFormCallback>,
            required: false,
            default(): void {
                /* do nothing */
            },
        },
    },
    data() {
        return {
            items: [
                { id: 1, name: "a", child: { name: "a1" } },
                { id: 2, name: "b", child: { name: "b1" } },
                { id: 3, name: "c", child: { name: "c1" } },
            ],
            createdEventData: null as null | ListItem,
            updatedEventData: null as null | ListItem,
            deletedEventData: null as null | ListItem,
            showBeforeValidationMessage: false,
        };
    },
    methods: {
        onCreatedEvent(item: ListItem): void {
            this.createdEventData = item;
        },
        onUpdatedEvent(item: ListItem): void {
            this.updatedEventData = item;
        },
        onDeletedEvent(item: ListItem): void {
            this.deletedEventData = item;
        },
    },
});

function createWrapper(
    templates = new Array<string>(),
    { stubs = [] as string[], props = {} } = {},
): VueWrapper<InstanceType<typeof TestComponent>> {
    const templateStart = `
        <f-crud-dataset v-model="items" :on-cancel="onCancel" :before-validation="beforeValidation" :before-submit="beforeSubmit" @created="onCreatedEvent" @updated="onUpdatedEvent" @deleted="onDeletedEvent">
            <template #default>
                <div id="cancelMessage"> Modalen har stängts </div>
                <table>
                    <tr>
                        <th> id </th>
                        <th> name </th>
                    </tr>
                    <tr v-for="item in items" :key="item.id">
                        <td> {{ item.id }} </td>
                        <td> {{ item.name }} </td>
                        <td> <f-crud-button id="modifyButton" action="modify" :item="item" label /> </td>
                        <td> <f-crud-button id="deleteButton" action="delete" :item="item" label /> </td>
                    </tr>
                </table>
            </template>
    `;
    const templateEnd = "</f-crud-dataset>";

    let templateMiddle = "";
    for (const t of templates) {
        templateMiddle += t;
    }
    const newComponent = defineComponent({
        template: templateStart.concat(templateMiddle).concat(templateEnd),
        extends: TestComponent,
    });

    return mount(newComponent, {
        global: {
            plugins: [ValidationPlugin],
            stubs: ["FIcon", ...stubs],
        },
        props,
        attachTo: createPlaceholderInDocument(),
    }) as VueWrapper<InstanceType<typeof TestComponent>>;
}

async function submitForm(wrapper: VueWrapper): Promise<void> {
    await wrapper.find(".modal form").trigger("submit");

    // hasFormErrors in FValidationForm
    await flushPromises(); // ValidationService.validateAllElements
    await wrapper.vm.$nextTick(); // FValidationForm.hasFormErrors.$nextTick
    await jest.runAllTimers(); // FValidationForm.hasFormErrors.setTimeout

    // hasFormErrors is called twice
    await flushPromises();
    await wrapper.vm.$nextTick();
    await jest.runAllTimers();

    await wrapper.vm.$nextTick(); // Remove modal from DOM
}

describe("snapshot", () => {
    it("should match snapshot", () => {
        const wrapper = createWrapper([ADD_TEMPLATE], {
            stubs: ["FFormModal", "FConfirmModal"],
        });
        expect(wrapper).toMatchSnapshot();
    });
});

it("should show add button when #add slot is present", () => {
    const wrapper = createWrapper([ADD_TEMPLATE], { stubs: ["FConfirmModal"] });
    expect(wrapper.find(".crud-dataset__add-button")).toMatchInlineSnapshot(`
        <button type="button" class="button button--tertiary crud-dataset__add-button">
          <f-icon-stub name="plus" library="f" class="button__icon"></f-icon-stub>Lägg till ny
        </button>
    `);
});

it("should show add modal when the add button is pressed", async () => {
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    const button = wrapper.find(".crud-dataset__add-button");
    await button.trigger("click");
    expect(wrapper.find("form > span").text()).toBe("add");
});

it("should show modify modal when the modify button is pressed", async () => {
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    const button = wrapper.find("#modifyButton");
    await button.trigger("click");
    expect(wrapper.find("form > span").text()).toBe("modify");
});

it("should show delete modal when the delete button is pressed", async () => {
    expect.assertions(2);
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );

    const confirmModal = wrapper.find("f-confirm-modal-stub");
    expect(confirmModal.attributes()["isopen"]).toBe("false");
    const button = wrapper.find("#deleteButton");
    await button.trigger("click");
    expect(confirmModal.attributes()["isopen"]).toBe("true");
});

it("should add item", async () => {
    const alertScreenReader = jest.spyOn(logic, "alertScreenReader");

    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    await wrapper.find(".crud-dataset__add-button").trigger("click");

    await wrapper.find("#id-input").setValue(4);
    await wrapper.find("#name-input").setValue("d");
    await submitForm(wrapper);

    const modifiedItem = wrapper.vm.$data.items[3];
    expect(modifiedItem.name).toBe("d");
    expect(alertScreenReader).toHaveBeenCalledWith("En rad har lagts till", {
        assertive: true,
    });
});

it("should modify items", async () => {
    const alertScreenReader = jest.spyOn(logic, "alertScreenReader");

    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    await wrapper.find("#modifyButton").trigger("click");

    await wrapper.find("#name-input").setValue("test");
    await wrapper.find("#child-name-input").setValue("test child");

    await submitForm(wrapper);

    const modifiedItem = wrapper.vm.$data.items[0];
    expect(modifiedItem.name).toBe("test");
    expect(modifiedItem.child.name).toBe("test child");
    expect(alertScreenReader).toHaveBeenCalledWith("Raden har ändrats", {
        assertive: true,
    });
});

it("should not modify original item on cancel", async () => {
    const wrapper = createWrapper([MODIFY_TEMPLATE], {
        stubs: ["FConfirmModal"],
    });
    const originalItem = wrapper.vm.$data.items[0];

    await wrapper.find("#modifyButton").trigger("click");

    await wrapper.find("#name-input").setValue("test");
    await wrapper.find("#child-name-input").setValue("test child");

    wrapper.find(".button-group .button--secondary").trigger("click");
    await jest.runAllTimers();
    await flushPromises();

    const updatedItem = wrapper.vm.$data.items[0];
    expect(originalItem).toEqual(updatedItem);
});

it("should emit created event when items are added", async () => {
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    await wrapper.find(".crud-dataset__add-button").trigger("click");

    await wrapper.find("#id-input").setValue(4);
    await wrapper.find("#name-input").setValue("d");
    await submitForm(wrapper);

    expect(wrapper.vm.$data.createdEventData?.id).toBe(4);
    expect(wrapper.vm.$data.createdEventData?.name).toBe("d");
});

it("should emit updated event when items are updated", async () => {
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );
    await wrapper.find("#modifyButton").trigger("click");

    await wrapper.find("#name-input").setValue("test");
    await submitForm(wrapper);

    expect(wrapper.vm.$data.updatedEventData?.id).toBe(1);
    expect(wrapper.vm.$data.updatedEventData?.name).toBe("test");
});

it("should call :onCancel after cancel", async () => {
    const wrapper = createWrapper(
        [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
        },
    );

    await wrapper.find(".crud-dataset__add-button").trigger("click");

    wrapper.find(".button-group .button--secondary").trigger("click");
    await jest.runAllTimers();
    await flushPromises();

    expect(wrapper.find("#cancelMessage").text()).toBe("Modalen har stängts");
});

describe("beforeSubmit", () => {
    it('should call "beforeSubmit" when it is provided and submit button clicked', async () => {
        const beforeSubmit = jest.fn();
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    beforeSubmit: beforeSubmit,
                },
            },
        );
        await wrapper.find(".crud-dataset__add-button").trigger("click");

        expect(beforeSubmit).toHaveBeenCalledTimes(0);

        await submitForm(wrapper);

        expect(beforeSubmit).toHaveBeenCalledTimes(1);
    });

    it("should not emit submit when beforeSubmit returns CANCEL", async () => {
        function onBeforeSubmit(): FFormModalAction {
            return FFormModalAction.CANCEL;
        }
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    beforeSubmit: onBeforeSubmit,
                },
            },
        );

        await wrapper.find(".crud-dataset__add-button").trigger("click");
        await submitForm(wrapper);

        expect(
            wrapper
                .findComponent(FCrudDataset)
                .findComponent(FFormModal)
                .emitted().submit,
        ).toBeFalsy();
    });

    it("should emit submit when beforeSubmit returns CONTINUE", async () => {
        function onBeforeSubmit(): FFormModalAction {
            return FFormModalAction.CONTINUE;
        }
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    beforeSubmit: onBeforeSubmit,
                },
            },
        );

        await wrapper.find(".crud-dataset__add-button").trigger("click");
        await submitForm(wrapper);

        expect(
            wrapper
                .findComponent(FCrudDataset)
                .findComponent(FFormModal)
                .emitted().submit,
        ).toBeTruthy();
    });

    it("should emit submit when beforeSubmit returns undefined", async () => {
        function onBeforeSubmit(): void {
            return;
        }
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    beforeSubmit: onBeforeSubmit,
                },
            },
        );

        await wrapper.find(".crud-dataset__add-button").trigger("click");
        await submitForm(wrapper);

        expect(
            wrapper
                .findComponent(FCrudDataset)
                .findComponent(FFormModal)
                .emitted().submit,
        ).toBeTruthy();
    });

    it("should emit submit when beforeSubmit props not set (default)", async () => {
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                },
            },
        );

        await wrapper.find(".crud-dataset__add-button").trigger("click");
        await submitForm(wrapper);

        expect(
            wrapper
                .findComponent(FCrudDataset)
                .findComponent(FFormModal)
                .emitted().submit,
        ).toBeTruthy();
    });
});

it("should call before validation", async () => {
    const beforeValidation = jest.fn();
    const wrapper = createWrapper(
        [ADD_TEMPLATE_REQUIRED, MODIFY_TEMPLATE, DELETE_TEMPLATE],
        {
            stubs: ["FConfirmModal"],
            props: {
                isOpen: true,
                beforeValidation: beforeValidation,
            },
        },
    );
    await wrapper.find(".crud-dataset__add-button").trigger("click");

    await wrapper.find("#id-input-required").setValue("");
    await wrapper.find("#name-input-required").setValue("");

    expect(beforeValidation).toHaveBeenCalledTimes(0);
    await submitForm(wrapper);
    expect(beforeValidation).toHaveBeenCalledTimes(1);
    expect(
        wrapper.findComponent(FCrudDataset).findComponent(FFormModal).emitted()
            .submit,
    ).toBeFalsy();
});

describe("onCancel", () => {
    it("should be called on cancel-event emitted", async () => {
        const onCancel = jest.fn();
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    onCancel: onCancel,
                },
            },
        );
        await wrapper.find(".crud-dataset__add-button").trigger("click");

        expect(onCancel).toHaveBeenCalledTimes(0);

        wrapper
            .findComponent(FCrudDataset)
            .findComponent(FFormModal)
            .vm.$emit("cancel");

        expect(onCancel).toHaveBeenCalledTimes(1);
    });

    it("should be called on close-button clicked", async () => {
        const onCancel = jest.fn();
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    onCancel: onCancel,
                },
            },
        );
        await wrapper.find(".crud-dataset__add-button").trigger("click");

        expect(onCancel).toHaveBeenCalledTimes(0);

        wrapper.find(".close-button").trigger("click");

        expect(onCancel).toHaveBeenCalledTimes(1);
    });

    it("should be called on cancel-button clicked", async () => {
        const onCancel = jest.fn();
        const wrapper = createWrapper(
            [ADD_TEMPLATE, MODIFY_TEMPLATE, DELETE_TEMPLATE],
            {
                stubs: ["FConfirmModal"],
                props: {
                    isOpen: true,
                    onCancel: onCancel,
                },
            },
        );
        await wrapper.find(".crud-dataset__add-button").trigger("click");

        expect(onCancel).toHaveBeenCalledTimes(0);

        wrapper.find(".button-group .button--secondary").trigger("click");

        expect(onCancel).toHaveBeenCalledTimes(1);
    });
});

describe("html-validate", () => {
    it("should require default slot", () => {
        expect("<f-crud-dataset></f-crud-dataset>").not.toHTMLValidate({
            message:
                '<f-crud-dataset> component requires slot "default" to be implemented',
        });
    });

    it("should html-validate", () => {
        expect(
            "<f-crud-dataset><template #default></template></f-crud-dataset>",
        ).toHTMLValidate();
    });
});
