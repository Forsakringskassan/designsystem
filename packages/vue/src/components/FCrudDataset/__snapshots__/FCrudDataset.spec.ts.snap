// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should match snapshot 1`] = `
<div class="crud-dataset">
  <div id="cancelMessage"> Modalen har stängts </div>
  <table>
    <tr>
      <th> id </th>
      <th> name </th>
    </tr>
    <tr>
      <td>1</td>
      <td>a</td>
      <td><button type="button" class="button button--small button--tertiary" id="modifyButton">
          <!--v-if-->
          <!--v-if-->Ändra
        </button></td>
      <td><button type="button" class="button button--small button--tertiary" id="deleteButton">
          <!--v-if-->
          <!--v-if-->Ta bort
        </button></td>
    </tr>
    <tr>
      <td>2</td>
      <td>b</td>
      <td><button type="button" class="button button--small button--tertiary" id="modifyButton">
          <!--v-if-->
          <!--v-if-->Ändra
        </button></td>
      <td><button type="button" class="button button--small button--tertiary" id="deleteButton">
          <!--v-if-->
          <!--v-if-->Ta bort
        </button></td>
    </tr>
    <tr>
      <td>3</td>
      <td>c</td>
      <td><button type="button" class="button button--small button--tertiary" id="modifyButton">
          <!--v-if-->
          <!--v-if-->Ändra
        </button></td>
      <td><button type="button" class="button button--small button--tertiary" id="deleteButton">
          <!--v-if-->
          <!--v-if-->Ta bort
        </button></td>
    </tr>
  </table>
  <div><button type="button" class="button button--tertiary crud-dataset__add-button">
      <f-icon-stub name="plus" library="f" class="button__icon"></f-icon-stub>Lägg till ny
    </button></div>
  <f-form-modal-stub fullscreen="true" isopen="false" size="" datatest="" useerrorlist="false" formid="fkui-vue-element-0001" ariaclosetext="Stäng" beforesubmit="[Function]" beforevalidation="[Function]" buttons="[object Object],[object Object]" value="[object Object]"></f-form-modal-stub>
  <f-confirm-modal-stub fullscreen="false" isopen="false" content="Brödtext" heading="Rubrik" size="" buttons="[object Object],[object Object]"></f-confirm-modal-stub>
</div>
`;
